### DISCUSSION TITLE
Run activity from another activity

### DISCUSSION THREAD

**Discussion by rosca-sabina** [elsa-workflows/elsa-core#5714] [Q&A] [answered]
Hello,

Is there a way to run another activity from a custom activity? I tried using method `context.ScheduleActivityAsync()`, but I'm having two issues:
- The `WriteLine` activity doesn't actually run.
- The completion callback runs immediately, not after `WriteLine` is executed.

Thanks in advance!

```csharp
public class RunAnotherActivity : Activity
{
    protected override async ValueTask ExecuteAsync(ActivityExecutionContext context)
    {
        var activityToSchedule = new WriteLine("Some message.");

        await context.ScheduleActivityAsync(activityToSchedule, Callback);
    }

    private async ValueTask Callback(ActivityCompletedContext context)
    {
        Console.WriteLine("Inner activity was executed.");

        await context.TargetContext.CompleteActivityAsync();
    }
}
```

**Comment by glime-ai** [comment]
To run another activity from within a custom activity and ensure that the completion callback runs after the scheduled activity has executed, you should make sure that the scheduled activity is properly awaited and that the callback is only invoked after the scheduled activity has completed.

In the code you provided, you are scheduling a `WriteLine` activity and providing a callback method. However, if the `WriteLine` activity doesn't run and the callback is invoked immediately, it could be due to the way the `WriteLine` activity is being scheduled or an issue with the workflow execution context.

Here are a few things you can check and try:

1. Ensure that the `WriteLine` activity is registered with the workflow runtime. If it's not registered, it won't be recognized and executed by the workflow engine.

2. Make sure that the `WriteLine` activity is properly configured to run within the context of the workflow. If there are any issues with the activity's configuration, it may not execute as expected.

3. Verify that the `WriteLine` activity is not being completed immediately for some reason. If the activity completes instantly, the callback will be invoked right away.

4. Check if there are any errors or exceptions being thrown during the execution of the `WriteLine` activity that could cause it to not run.

5. Review the workflow logs to see if there are any messages that could indicate why the `WriteLine` activity is not running.

6. If you are using Elsa 3.0, you can refer to the documentation or community discussions for examples of how to schedule activities and handle their completion. For instance, the GitHub discussions and issues related to Elsa workflows might provide insights or similar scenarios that can help you troubleshoot the problem.

If after checking these points the issue still persists, you may want to provide more details about your workflow configuration and the context in which the `RunAnotherActivity` is being executed. This information could be crucial to understanding the root cause of the problem.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Comment by rosca-sabina** [comment, marked as answer]
I found two methods to achieve the desired behavior:
- If you want to keep `RunAnotherActivity` of type `Activity` or `CodeActivity`, the inner activity (in our example `WriteLine`) needs to be declared in the `WorkflowBase.Build()` method of your workflow. If you declare the inner activity directly inside `RunAnotherActivity.ExecuteAsync()` it doesn't work.
- You can make `RunAnotherActivity` a Composite activity by extending `Composite` instead of `Activity`, and there are examples in the Elsa samples of how to do this: https://github.com/elsa-workflows/elsa-samples/tree/main/src/console/Elsa.Samples.ConsoleApp.Composition
